File: Stack/173.binary-search-tree-iterator.go
package leetcode

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type BSTIterator struct {
	st []*TreeNode
}

func Constructor(root *TreeNode) BSTIterator {
	st := []*TreeNode{}
	for root != nil {
		st = append(st, root)
		root = root.Left
	}
	return BSTIterator{
		st: st,
	}
}

func (this *BSTIterator) Next() int {
	res := this.st[len(this.st)-1]
	this.st = this.st[:len(this.st)-1]
	cur := res.Right
	for cur != nil {
		this.st = append(this.st, cur)
		cur = cur.Left
	}
	return res.Val
}

func (this *BSTIterator) HasNext() bool {
	return len(this.st) > 0
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */
// @leet end

File: Stack/173.binary-search-tree-iterator.py
from typing import Optional


# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.st = []
        while root:
            self.st.append(root)
            root = root.left

    def next(self) -> int:
        res = self.st.pop()
        cur = res.right
        while cur:
            self.st.append(cur)
            cur = cur.left
        return res.val

    def hasNext(self) -> bool:
        return len(self.st) > 0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
# @leet end


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator2:
    def __init__(self, root: Optional[TreeNode]):
        self.st = []
        while root:
            self.st.append(root)
            root = root.left

    def next(self) -> int:
        res = self.st.pop()
        cur = res.right
        while cur:
            self.st.append(cur)
            cur = cur.left

        return res.val

    def hasNext(self) -> bool:
        return len(self.st) > 0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
